### ISSUE TITLE
Narrowed types from user defined type guards are lost in subsequent method calls

### ISSUE THREAD

**Issue by OliverJAsh** [ReactiveX/IxJS#162]
**IxJS version:** 2.3.1

**Code to reproduce:**
``` ts
// no error, good!
const xs2: Ix.Iterable<number> = Ix.Iterable
    .from(['foo', 1])
    .filter((x): x is number => typeof x === 'number')

// no error, good!
const xs2b: Ix.Iterable<number> = xs2.map(x => x);

// unexpected error, bad!
const xs3: Ix.Iterable<number> = Ix.Iterable
    .from(['foo', 1])
    .filter((x): x is number => typeof x === 'number')
    .map(x => x)
```

Related https://github.com/ReactiveX/IxJS/issues/44

**Comment by trxcllnt** [comment]
@OliverJAsh ah, so I think this is a typescript bug. Check this out:

```typescript
// no error
const arr1: Array<number> = ['foo', 1]
  .filter((x): x is number => typeof x === 'number');

/*
[ts]
Type '(string | number)[]' is not assignable to type 'number[]'.
  Type 'string | number' is not assignable to type 'number'.
  Type 'string' is not assignable to type 'number'.
*/
const arr2: Array<number> = ['foo', 1]
  .filter((x): x is number => typeof x === 'number')
  .map((x) => x);

// no error
const arr3: Array<number> = ['foo', 1]
  .filter((x: any): x is number => typeof x === 'number')
  .map((x) => x);
```

It looks like adding an `any` type annotation to the filter predicate argument fakes typescript out somehow. Doing the same with Iterable yields roughly the same behavior:

```typescript
import { Iterable as IterableX } from "ix";

// no error
const itr1: IterableX<number> = IterableX.from(['foo', 1])
  .filter((x): x is number => typeof x === 'number');

/*
[ts]
Type 'IterableX<string | number>' is not assignable to type 'IterableX<number>'.
  Type 'string | number' is not assignable to type 'number'.
  Type 'string' is not assignable to type 'number'.
 */
const itr2: IterableX<number> = IterableX.from(['foo', 1])
  .filter((x): x is number => typeof x === 'number')
  .map(x => x);

// no error
const itr3: IterableX<number> = IterableX.from(['foo', 1])
  .filter((x: any): x is number => typeof x === 'number')
  .map(x => x);
```

And here's something else that's strange; the source Array's generic type is preserved, while the source Iterable's is erased:
<img width="541" alt="screen shot 2017-11-21 at 1 50 00 am" src="https://user-images.githubusercontent.com/178183/33065735-693d5732-ce5e-11e7-9e1a-77306fe9a3f2.png">
<img width="544" alt="screen shot 2017-11-21 at 1 50 28 am" src="https://user-images.githubusercontent.com/178183/33065742-69f22432-ce5e-11e7-98d2-0d79fb16f733.png">

It looks like TS is inferring T from the filter predicate's `value: T` argument, rather than `this: Iterable<T>`. Changing `value: T` to `value: any` for the [filter type-guard signature](https://github.com/ReactiveX/IxJS/blob/master/src/add/iterable-operators/filter.ts#L9) fixes this problem, but I worry we'd be defeating the compiler if this were to be fixed in a future release.

**Comment by OliverJAsh** [comment]
@trxcllnt Good investigation! I opened an issue on TypeScript with regards to this. See https://github.com/Microsoft/TypeScript/issues/20186.

**Comment by trxcllnt** [comment]
@OliverJAsh also, looks like defining the predicate before chaining also fixes it:
```typescript
function isNumber(x: string | number): x is number {
  return typeof x === 'number';
}
// no error
const arr: Array<number> = ['foo', 1].filter(isNumber).map((x) => x);
// no error
const itr3: IterableX<number> = IterableX.from(['foo', 1]).filter(isNumber).map(x => x);
```

**Comment by trxcllnt** [comment]
(cross-posting from https://github.com/Microsoft/TypeScript/issues/20186)

@OliverJAsh based on https://github.com/Microsoft/TypeScript/issues/19640#issuecomment-341122227 it looks like this may be fixed by #17600, as this form also works in 2.6.1:

```typescript
// no error
const arr: Array<number> = ['foo', 1]
  .filter<number>((x): x is number => typeof x === 'number')
  .map((x) => x);
```
